File: app/src/main/java/com/controlmoblie/llm/CommandTemplates.java
package com.controlmoblie.llm;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

public class CommandTemplates {

    private static class Template {
        final List<String> keywords;
        final java.util.function.Function<String, Map<String, String>> extract;
        final java.util.function.Function<Map<String, String>, String> buildJson;

        Template(List<String> keywords,
                 java.util.function.Function<String, Map<String, String>> extract,
                 java.util.function.Function<Map<String, String>, String> buildJson) {
            this.keywords = keywords;
            this.extract = extract;
            this.buildJson = buildJson;
        }
    }

    private static final List<Template> TEMPLATES = Arrays.asList(
        new Template(
            Arrays.asList("发消息给", "发消息"),
            CommandTemplates::extractSendMessage,
            p -> buildSequenceJson(
                waitStep(500), clickStep("通讯录"), waitStep(300),
                clickStep(p.getOrDefault("target", "")), waitStep(300),
                typeStep(p.getOrDefault("text", "")), waitStep(300),
                clickStep("发送")
            )
        ),
        new Template(
            Arrays.asList("看朋友圈"),
            input -> new HashMap<>(),
            p -> "{\"action\":\"open_wechat_page\",\"page\":\"moments\"}"
        ),
        new Template(
            Arrays.asList("发朋友圈"),
            input -> {
                String text = input.substring(input.indexOf("发朋友圈") + 4).trim();
                if (text.startsWith("说")) text = text.substring(1).trim();
                if (text.isEmpty()) return null;
                Map<String, String> map = new HashMap<>();
                map.put("text", text);
                return map;
            },
            p -> buildSequenceJson(
                waitStep(500), clickStep("发现"), waitStep(300),
                clickStep("朋友圈"), waitStep(500),
                clickStep("拍照分享"), waitStep(300),
                typeStep(p.getOrDefault("text", "")), waitStep(300),
                clickStep("发表")
            )
        ),
        new Template(
            Arrays.asList("打开小程序"),
            input -> {
                String name = input.substring(input.indexOf("打开小程序") + 4).trim();
                if (name.isEmpty()) return null;
                Map<String, String> map = new HashMap<>();
                map.put("name", name);
                return map;
            },
            p -> buildSequenceJson(
                waitStep(500),
                scrollStep("down", "short"),
                waitStep(300), clickStep(p.getOrDefault("name", ""))
            )
        ),
        new Template(
            Arrays.asList("搜索公众号"),
            input -> {
                String name = input.substring(input.indexOf("搜索公众号") + 4).trim();
                if (name.isEmpty()) return null;
                Map<String, String> map = new HashMap<>();
                map.put("name", name);
                return map;
            },
            p -> buildSequenceJson(
                waitStep(500), clickStep("通讯录"), waitStep(300),
                clickStep("公众号"), waitStep(300),
                typeStep(p.getOrDefault("name", "")), waitStep(500),
                clickStep(p.getOrDefault("name", ""))
            )
        ),
        new Template(
            Arrays.asList("的朋友圈"),
            input -> {
                int idx = input.indexOf("的朋友圈");
                if (idx < 0) return null;
                String target = input.substring(0, idx).trim();
                if (target.startsWith("看")) target = target.substring(1).trim();
                if (target.isEmpty()) return null;
                Map<String, String> map = new HashMap<>();
                map.put("target", target);
                return map;
            },
            p -> buildSequenceJson(
                waitStep(500), clickStep("通讯录"), waitStep(300),
                clickStep(p.getOrDefault("target", "")), waitStep(300),
                clickStep("朋友圈")
            )
        )
    );

    private static Map<String, String> extractSendMessage(String input) {
        if (input.contains("发消息给") && input.contains("说")) {
            String after = input.substring(input.indexOf("发消息给") + 4).trim();
            int sayIdx = after.indexOf("说");
            if (sayIdx < 0) return null;
            String target = after.substring(0, sayIdx).trim();
            String text = after.substring(sayIdx + 1).trim();
            if (target.isEmpty() || text.isEmpty()) return null;
            Map<String, String> map = new HashMap<>();
            map.put("target", target);
            map.put("text", text);
            return map;
        }
        if (input.contains("给") && input.contains("发消息")) {
            String afterGei = input.substring(input.indexOf("给") + 1).trim();
            int msgIdx = afterGei.indexOf("发消息");
            if (msgIdx < 0) return null;
            String target = afterGei.substring(0, msgIdx).trim();
            String text = afterGei.substring(msgIdx + 3).trim();
            if (target.isEmpty() || text.isEmpty()) return null;
            Map<String, String> map = new HashMap<>();
            map.put("target", target);
            map.put("text", text);
            return map;
        }
        return null;
    }

    public static String match(String userText) {
        for (Template tpl : TEMPLATES) {
            boolean matched = false;
            for (String kw : tpl.keywords) {
                if (userText.contains(kw)) {
                    matched = true;
                    break;
                }
            }
            if (matched) {
                Map<String, String> params = tpl.extract.apply(userText);
                if (params == null) continue;
                return tpl.buildJson.apply(params);
            }
        }
        return null;
    }

    private static String clickStep(String target) {
        return "{\"action\":\"click\",\"target\":\"" + escapeJson(target) + "\"}";
    }

    private static String typeStep(String text) {
        return "{\"action\":\"type\",\"text\":\"" + escapeJson(text) + "\"}";
    }

    private static String waitStep(long ms) {
        return "{\"action\":\"wait\",\"ms\":" + ms + "}";
    }

    private static String scrollStep(String direction, String distance) {
        return "{\"action\":\"scroll\",\"direction\":\"" + escapeJson(direction)
            + "\",\"distance\":\"" + escapeJson(distance) + "\"}";
    }

    private static String escapeJson(String value) {
        if (value == null) return "";
        return value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r");
    }

    private static String buildSequenceJson(String... steps) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"action\":\"sequence\",\"steps\":[");
        for (int i = 0; i < steps.length; i++) {
            if (i > 0) sb.append(",");
            sb.append(steps[i]);
        }
        sb.append("]}");
        return sb.toString();
    }
}

File: app/src/main/java/com/controlmoblie/llm/CommandTemplates.kt
package com.controlmoblie.llm

object CommandTemplates {

    private data class Template(
        val keywords: List<String>,
        val extract: (String) -> Map<String, String>?,
        val buildJson: (Map<String, String>) -> String
    )

    private val templates: List<Template> = listOf(
        Template(
            keywords = listOf("发消息给", "发消息"),
            extract = { input -> extractSendMessage(input) },
            buildJson = { p ->
                buildSequenceJson(
                    step("wait", 500), step("click", "通讯录"), step("wait", 300),
                    step("click", p["target"] ?: ""), step("wait", 300),
                    step("type", p["text"] ?: ""), step("wait", 300),
                    step("click", "发送"),
                )
            }
        ),
        Template(
            keywords = listOf("看朋友圈"),
            extract = { emptyMap() },
            buildJson = {
                buildSequenceJson(
                    step("wait", 500), step("click", "发现"), step("wait", 300),
                    step("click", "朋友圈"),
                )
            }
        ),
        Template(
            keywords = listOf("发朋友圈"),
            extract = { input ->
                val text = input.substringAfter("发朋友圈").trim().removePrefix("说").trim()
                if (text.isBlank()) null else mapOf("text" to text)
            },
            buildJson = { p ->
                buildSequenceJson(
                    step("wait", 500), step("click", "发现"), step("wait", 300),
                    step("click", "朋友圈"), step("wait", 500),
                    step("click", "拍照分享"), step("wait", 300),
                    step("type", p["text"] ?: ""), step("wait", 300),
                    step("click", "发表"),
                )
            }
        ),
        Template(
            keywords = listOf("打开小程序"),
            extract = { input ->
                val name = input.substringAfter("打开小程序").trim()
                if (name.isBlank()) null else mapOf("name" to name)
            },
            buildJson = { p ->
                buildSequenceJson(
                    step("wait", 500),
                    step("scroll", mapOf("direction" to "down", "distance" to "short")),
                    step("wait", 300), step("click", p["name"] ?: ""),
                )
            }
        ),
        Template(
            keywords = listOf("搜索公众号"),
            extract = { input ->
                val name = input.substringAfter("搜索公众号").trim()
                if (name.isBlank()) null else mapOf("name" to name)
            },
            buildJson = { p ->
                buildSequenceJson(
                    step("wait", 500), step("click", "通讯录"), step("wait", 300),
                    step("click", "公众号"), step("wait", 300),
                    step("type", p["name"] ?: ""), step("wait", 500),
                    step("click", p["name"] ?: ""),
                )
            }
        ),
        Template(
            keywords = listOf("的朋友圈"),
            extract = { input ->
                val idx = input.indexOf("的朋友圈")
                if (idx < 0) null
                else {
                    val target = input.substring(0, idx).trim().removePrefix("看").trim()
                    if (target.isBlank()) null else mapOf("target" to target)
                }
            },
            buildJson = { p ->
                buildSequenceJson(
                    step("wait", 500), step("click", "通讯录"), step("wait", 300),
                    step("click", p["target"] ?: ""), step("wait", 300),
                    step("click", "朋友圈"),
                )
            }
        ),
    )

    private fun extractSendMessage(input: String): Map<String, String>? {
        if (input.contains("发消息给") && input.contains("说")) {
            val after = input.substringAfter("发消息给").trim()
            val sayIdx = after.indexOf("说")
            if (sayIdx < 0) return null
            val target = after.substring(0, sayIdx).trim()
            val text = after.substring(sayIdx + 1).trim()
            if (target.isBlank() || text.isBlank()) return null
            return mapOf("target" to target, "text" to text)
        }
        if (input.contains("给") && input.contains("发消息")) {
            val afterGei = input.substringAfter("给").trim()
            val msgIdx = afterGei.indexOf("发消息")
            if (msgIdx < 0) return null
            val target = afterGei.substring(0, msgIdx).trim()
            val text = afterGei.substring(msgIdx + 3).trim()
            if (target.isBlank() || text.isBlank()) return null
            return mapOf("target" to target, "text" to text)
        }
        return null
    }

    fun match(userText: String): String? {
        for (tpl in templates) {
            val matched = tpl.keywords.any { userText.contains(it) }
            if (matched) {
                val params = tpl.extract(userText) ?: continue
                return tpl.buildJson(params)
            }
        }
        return null
    }

    private fun step(type: String, target: String): String =
        """{"action":"$type","target":"$target"}"""

    private fun step(type: String, ms: Long): String =
        """{"action":"$type","ms":$ms}"""

    private fun step(type: String, props: Map<String, String>): String {
        val inner = props.entries.joinToString(",") { """"${it.key}":"${it.value}"""" }
        return """{"action":"$type",$inner}"""
    }

    private fun buildSequenceJson(vararg steps: String): String {
        val stepsJson = steps.joinToString(",")
        return """{"action":"sequence","steps":[$stepsJson]}"""
    }
}
